feat: 홈 화면 구현

nil-ian opened the pull request:
❗️ 이슈 번호
Closes #13 

### 📝 작업 유형

- [x] 신규 기능 추가

### 📙 작업 내역

- 디자인 에셋 및 리소스
  - 홈 화면에 필요한 디자인 에셋 추가
  - Color 생성시 alpha값도 함께 전달받도록 변경
- BaseClass
  - BaseViewController: gradient 배경 적용하도록 변경
  - BaseCollectionVIewCell: init시 UI 관련 함수 호출하도록 변경
- Appearance
  - NavigationBar, TabBar Appearance 적용
- Home
  - Domain Layer 구현(Entity, Service Interface, UseCase)
  - Service Layer 구현(VideoPostAPI, VideoPostService)
  - UI
      - VIdeoPlayerView 구현
      - VideoPostCell 구현
      - HomeViewController, HomeReactor 구현 

<br>
<br>

### 💡 참고사항 (없다면 지워도 됩니다!)

- 추후 API가 나오면 연동 예정

nil-ian committed:
design: 디자인 리소스 에셋 추가

nil-ian committed:
fix: 불필요한 파일 제거 및 수정

HomeCoordinator 파일 제거 및 AppCoordinator 수정

nil-ian committed:
feat: 익스텐션 기능 수정 및 추가

- createColor시 alpha값도 함꼐 변경할 수 있도록 수정
- UIView의 corenrRadius 익스텐션 추가
- Array subscript시 Safe한 방식 제공

nil-ian committed:
feat: #13 - HomeScene Factory 적용

nil-ian committed:
fix: BaseVC, BaseCell 수정

- BaseVC: 배경 색상 수정
- BaseCell: init시 UI 관련 함수 호출하도록 수정

nil-ian committed:
design: UIAppearance 수정

- NavigationBar 및 TabBar Appearance 디자인에 맞게 수정

nil-ian committed:
feat: #13 - VIdeoPlayerView 구현

nil-ian committed:
feat: #13 - Domain Layer 구현

- Entity(VideoPost, Shorts) 정의
- Interface(VideoPostServiceType) 정의
- UseCase(FetchVideoPosts, ExclameVideoPost, LikeVideoPost) 정의
- 의존성 Assemble

nil-ian committed:
feat: #13 - Service Layer 구현

- VIdeoPostAPI 정의
- VideoPostService 구현
  - 현재는 더미데이터를 반환중, 추후 API 나오면 연동 예정
- 의존성 Assemble

nil-ian committed:
feat: #13 - VideoPostCell 구현

nil-ian committed:
feat: #13 - HomeReactor 구현

nil-ian committed:
feat: #13 - HomeViewController 구현

nil-ian committed:
feat: #13 - Paging 기능 추가

- didEndDragging 시점에 페이징 호출

Hansangjin98 commented on Targets/DesignKit/Sources/Views/VIdeoPlayerView.swift:
```diff
@@ -0,0 +1,42 @@
+//
+//  VIdeoPlayerView.swift
+//  DesignKit
```
이 파일 이름 오타!

Hansangjin98 commented on Targets/Feature/Sources/Common/Classes/SceneFactory.swift:
```diff
@@ -50,33 +55,88 @@ final class SceneFactory: SceneFactoryType {
       return UIViewController()
+
     case .home:
-      let homeViewController: HomeViewController = .init(nibName: nil, bundle: nil)
+      let coordinator = injector.resolve(AppCoordinatorType.self)
+      let homeUseCase = injector.resolve(HomeUseCaseType.self)
+      let homeViewController: HomeViewController = .init(
+        reactor: .init(
+          dependency: .init(
+            coordinaotr: coordinator,
+            useCase: homeUseCase
+          )
+        )
+      )
```
탭바에 들어가는 Scene들(home, videoUpload, profile)은 create 메서드에서 생성할 때 중복 코드가 발생하게 되는데 이 부분 개선해 보면 좋을 것 같아!
탭바에 포함 안된 채로 단독으로 create될 경우가 없으면 아예 switch case에서 빼도 괜찮을 듯!

nil-ian replied:
혹시나 사용될까봐 생성해두었는데 일단은 제거하는거로!

Hansangjin98 commented on Targets/Feature/Sources/Home/View/HomeViewController.swift:
```diff
@@ -8,13 +8,210 @@
+    }
+  }
+}
+
+// MARK: - UICollectionViewDelegate
+extension HomeViewController: UICollectionViewDelegate {}
```
delegate 메서드 쓰는 거 없으면 채택할 필요 없을 것 같아~
80번째 줄이랑 같이 제거해도 될 듯!

nil-ian replied:
습관 무섭다ㅠㅠㅋㅋㅋ 굳굳!!

Hansangjin98 commented on Targets/Feature/Sources/Home/View/HomeViewController.swift:
```diff
@@ -8,13 +8,210 @@
+      var offset = contentOffset.pointee
+      let index = round((offset.x + owner.collectionView.contentInset.left) / cellWidth)
+
+      if index > CGFloat(owner.currentIndex) {
+        owner.currentIndex += 1
+      } else if index < CGFloat(owner.currentIndex) {
+        if owner.currentIndex != 0 {
+          owner.currentIndex -= 1
+        }
+      }
```
가독성이 쪼끔 아쉬운데 개선해 볼 수 있을 것 같아!!

Hansangjin98 reviewed (commented):
굿굿 고생 많았어~!!!

nil-ian committed:
fix: 불필요한 파일 제거

- Profile, VIdeoUpload Coordinator 제거

nil-ian committed:
feat: #13 - 리뷰내용 반영

- 오타 수정(VIdeoPlayerVIew -> VideoPlayerView)
- SceneFactory 내 Tab 구성 Scene 제거
- if문 컨디션 뎁스 줄이기
